Boost VRP ListSelector’s Scalability with Nearby (#581)
kiegroup/optaplanner-website: optaplanner-website-root/content/blog/2023/04/05/BoostVRPListSelectorScalability.adoc

= Boosting list variable scalability with nearby selection
dupliaka

> 2023-04-05

Reviewer: Let's not forget to update the date as well as the directory structure when the blogpost is ready for publishing.

To create candidate solution solver picks Customers one by one and selects to whom of Vehicle's Customer list it should belong and in which order.
The list is called https://www.optaplanner.org/docs/optaplanner/latest/planner-configuration/planner-configuration.html#planningListVariable[PlanningListVariable] and the set of changes from solution A to solution B is called a https://www.optaplanner.org/docs/optaplanner/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#moveAndNeighborhoodSelectionIntroduction[Move].
Optimization algorithms rely on Moves to shift between solutions.
This means they must master https://www.optaplanner.org/docs/optaplanner/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#whatIsAMoveSelector[move selection] - picking the best moves quickly and effectively.

VRP move selection faces a significant challenge when it comes to scalability.
With each additional node in the problem, the complexity of the algorithm increases exponentially, making it difficult to find feasible solutions within a reasonable amount of time.

== How to address this challenge?

> https://www.optaplanner.org/docs/optaplanner/latest/move-and-neighborhood-selection/move-and-neighborhood-selection.html#nearbySelection[Nearby selection (NS)]  has recently been implemented for the `PlanningListVariable` as a tool to reduce the amount of nodes that need to be considered during the optimization process.

Author: 3081c46e

Imagine you're a delivery company trying to deliver packages to different cities with the minimum driving time spent.
Suppose you need to deliver a package from Berlin to Paris.
Does it make sense to put a stop in Tokyo in the middle?

== How does NS improve the calculation?

Let's compare the solver progress over time for plain Late Acceptance (LA) solver and LA with the nearby move selector (LANS).
On the graphs below you can see the https://github.com/kiegroup/optaplanner/blob/main/optaplanner-examples/src/main/resources/org/optaplanner/examples/vehiclerouting/optional/benchmark/vehicleRoutingBenchmarkConfigListNearby.xml[optaplanner-benchmark] results for different amount of samples.
You can find out more about optaplanner-benchmarks https://www.optaplanner.org/docs/optaplanner/latest/benchmarking-and-tweaking/benchmarking-and-tweaking.html[here].

== Execution time

How fast each algorithm improves the solution score?

Graph below represents the speed of score calculation for different selector options for a certain amount of samples.

> The x-axis represents the size of datasets from 50 customers with 10 vehicles to 1000 customers with 20 vehicles while the y-axis represents the ratio of score calculation speed = best score/time spent.

Author: 3013164e

|Late Acceptance                      >|35,393.00/s >|37,334.00/s >|31,667.00/s >|30,282.00/s
|LV Nearby Parabolic Distribution 10  >|40,428.00/s >|45,059.00/s >|32,708.00/s >|32,036.00/s
|LV Nearby Parabolic Distribution 20  >|38,567.00/s >|40,552.00/s >|34,418.00/s >|37,188.00/s
|LV Nearby Parabolic Distribution 40  >|37,551.00/s >|39,833.00/s >|33,558.00/s >|35,125.00/s
|LV Nearby Parabolic Distribution 80  >|37,428.00/s >|35,625.00/s >|30,523.00/s >|36,060.00/s
|LV Nearby Parabolic Distribution 160 >|41,197.00/s >|38,086.00/s >|33,724.00/s >|34,451.00/s
|===

== Solution quality

Despite the fact that NS improves the speed of calculation, it could potentially decrease the quality of the best solution.

image::bestScoreNB.png[How NS affects scores]

The image above represents the overall best score archived (y-axis) with different move selectors and datasets for the same amount of time (5 minutes).
The less is better.

You may see how much the NS improves the score calculation while the dataset increases.
For the biggest dataset any LANS variation gain twice better score then LA.

From the other side look at the LV Nearby 10 bars.

> Despite being the fastest solution of all with the small dataset, it somehow dramatically loose the score gain for the case of 500 customers and 20 vehicles.

Reviewer: ```suggestion
Despite being the fastest solution of all with the small dataset, it somehow dramatically lose the score gain for the case of 500 customers and 20 vehicles.
```

> The graph below shows how in this case best score (y-axis) changes in time during the calculation . The less is better.

Reviewer: Again, I would prefer to have the axes' descriptions directly in the graph.

If you show the score as a single-fold number, what does it actually mean? Is it the simple score, the hard score or the soft score? If the benchmark used the hardsoft score, ideally, let's show it in the graph as 0hard/-20000soft. Or at least let's make clear we talk about soft score given all solutions we'got were feasible.

Author: 38291a34

image::belgium-n500-k20BestScoreStatisticLevel.png[How local optimum affects the score]

You can clearly see that at some point due to the small distribution size it fails to reach better solution and not able to get out of the local optimum providing the same result as LA.

== Scalability

> image::bestScoreScalabilitySummaryLevel1.png[Scalability of NS solver]

Reviewer: I like the graph. Especially the description of axes in it. :-)

Author: I cannot recognize the irony here)) I removed the description of axes 38291a34

Reviewer: No irony intended; when I was writing the comment, this graph already had the correct axes description.

The graph is showing the performance of different selector options for optimizing the delivery of goods for a certain number of customers and vehicles.
The x-axis represents the size of the dataset, ranging from small (less than 500 samples) to large (1 million samples), while the y-axis represents the ratio of the best score achieved by each selector option.

> The graph is suggesting that for small datasets with a low number of customers (less than 500), the choice of selector option does not have a significant impact on the best score achieved.

Reviewer: I am a bit lost in what "choice of selector option" means; given the context, I assume you talk about nearby selection size parameter.

Author: No, there is different selection group: lika LA default - the one selector and yes, the selector with different sizes is also the option

Author: addressed 82009f51

> = Boost VRP ListSelector's Scalability with Nearby Selector

Reviewer 2: ```suggestion
= Boosting list variable scalability with nearby selection
```

> = Boost VRP ListSelector's Scalability with Nearby Selector

Author: Ok, how about :"Scaling Up Vehicle Routing with Planning List Variable and Nearby Selector"

dupliaka
2023-04-05
:page-interpolate: true
:jbake-type: post
:jbake-tags: insight, design, vehicle routing. move selector
:jbake-social_media_share_image: nearby-selection.png

As a solver of optimization problems, the VRP ListSelector faces a significant challenge when it comes to scalability.

[quote]
____

> "`With each additional node in the problem, the complexity of the algorithm increases exponentially, making it difficult to find feasible solutions within a reasonable amount of time.`"

Reviewer 2: Who are we quoting?

Author: Nobody. I will fix that

Author: 9ba65a3b

____

== How to address this challenge?

> `NearbyMoveSelector` has been implemented as a tool to reduce the amount of nodes that need to be considered during the optimization process. By pre-calculating a distance matrix and using different distributions to limit the number of moves that need to be considered, the nearbyMoveSelector improves the scalability of the algorithm, making it possible to tackle larger and more complex problems.

Reviewer 2: ```suggestion
Nearby selection has recently been implemented for the `PlanningListVariable` as a tool to reduce the amount of nodes that need to be considered during the optimization process. By pre-calculating a distance matrix and using different distributions to limit the number of moves that need to be considered, this selector improves the scalability of the algorithm, making it possible to tackle larger and more complex problems.
```

But what is the underlying analysis behind the success of the `NearbyMoveSelector`, and where might bottlenecks arise in the optimization process?

> At its core, the `NearbyMoveSelector` is a tool for reducing the complexity of the optimization process. By limiting the number of moves that need to be considered, it reduces the number of comparisons required, ultimately leading to a reduction in execution time. This is achieved by pre-calculating a distance matrix, which allows the algorithm to quickly identify the closest nodes and *focus its attention on those that are most relevant*.

Author: 9ba65a3b

== How that improves the calculation?

> Let's compare the solver progress over time with `NearbyMoveSelector` and the Default configuration. On the graph below you can see the _optaplanner-benchmarks/.../vehicleRoutingBenchmarkConfigListNearBy.xml_ benchmark results for different types of distributions with *50 Customers* and *10 Vehicles*.  The maximum sampling for the Move Selector in this case is 60 (sum of Planning Values(50) and Planning Entities(10)) that is why the Distribution Size maximum was picked with  https://youexec.com/book-summaries/algorithms-to-live-by[37% of the amount]. By the dynamic of the curve you can see that the PARABOLIC_DESTRIBUTION `NearbyMoveSelector` was the fastest and it overtook `LateAsseptance` being twice as fast.

Author: Agree let me provide a link to it when the PR will be merged

https://github.com/kiegroup/optaplanner/pull/2747#issuecomment-1497608737

image::listNearByBenchmark.png[How NearByMoveSelector improves score calculation speed]

== Potential bottlenecks?

> However, there are potential bottlenecks that can arise while using of the `NearbyMoveSelector`. For example, if the distribution used to limit the number of moves is too restrictive, it may exclude nodes that could be part of a feasible solution. Similarly, if the amount of supplied moves would be sufficiently large, it can provide a better score while scaling poorly.

Reviewer 2: ```suggestion
There are potential bottlenecks that can arise while using nearby selection. For example, if the distribution used to limit the number of moves is too restrictive, it may exclude nodes that could be part of a feasible solution. Similarly, if the amount of supplied moves were sufficiently large, it can provide a better score while scaling poorly.
```

Also - "provide a better score while scaling poorly"? This makes no sense to me. Seems like contradicting itself.

image::listNearByBenchmarkScore.png[How NearByMoveSelector improves score calculation speed]

> Here we can see that the BLOCK_DESTRIBUTION won with the lowest score _-21 727m_, while the fastest was only _24m_ longer and the worst - LA is longer by 2% of the best solution.  Beside the fact that the score, calculated with `NearbyMoveSelector` was better you can notice that the best score was brought by the BLOCK_DESTRIBUTION that is because the probability for any sample from 37% range of moves is the same, that reflects the 37% rule.

Author: I shared the link https://github.com/dupliaka/optaplanner-website/blob/0141046bd29295092c87b4fdf5ba730f74c090c6/optaplanner-website-root/content/blog/2023/04/05/BoostVRPListSelectorScalability.adoc#how-that-improves-the-calculation

I was need to setup a limit and keep it simply the same for the results. Also I wanted to show the improtance of the sensitivity of the parameters. https://youexec.com/book-summaries/algorithms-to-live-by

> Here we can see that the BLOCK_DESTRIBUTION won with the lowest score _-21 727m_, while the fastest was only _24m_ longer and the worst - LA is longer by 2% of the best solution.  Beside the fact that the score, calculated with `NearbyMoveSelector` was better you can notice that the best score was brought by the BLOCK_DESTRIBUTION that is because the probability for any sample from 37% range of moves is the same, that reflects the 37% rule.

Author: I found it is as an amazing example of working the rule in the real life and it is really good

> So, as we can see it is important to carefully select the distribution and parameters used by the `NearbyMoveSelector`. This requires a deep understanding of the problem being solved, as well as careful attention to detail and a willingness to iterate and refine the optimization process over time.

Reviewer 2: ```suggestion
So, as we can see it is important to carefully select the distribution and parameters used by nearby selection. This requires deep understanding of the problem being solved, as well as careful attention to detail and willingness to iterate and refine the optimization process over time.
```

== Conclusion

> Overall, the success of the `NearbyMoveSelector` demonstrates the power of carefully designed tools for reducing the complexity of optimization problems. By leveraging pre-calculated data and strategic algorithms, it is possible to overcome the challenges posed by scalability and tackle ever more complex problems in the future.

Reviewer 2: Far too many words, far too little actual content.

= Boosting list variable scalability with nearby selection
dupliaka
2023-04-05
:page-interpolate: true
:jbake-type: post
:jbake-tags: insight, design, vehicle routing. move selector
:jbake-social_media_share_image: nearby-selection.png

> As a solver of optimization problems, the VRP ListSelector faces a significant challenge when it comes to scalability.

Reviewer: Unfortunately, there has not been any article about the list planning variable itself. Thus, this article builds upon something the user cannot be familiar with.

I cannot ask you to write a standalone blog post about the feature, but could you at least provide an introduction to it, referring to the documentation for those who want to learn more?

Nearby selection has been there for the chained model for a long already; this is about closing the gaps between the chained model and the list planning variable.

Imagine you're a delivery company trying to deliver packages to different cities with the minimum driving time spent.
Suppose you need to deliver a package from Berlin to Paris. Does it make sense to put a stop in Tokyo in the middle?

== How does nearby selection improve the calculation?

Let's compare the solver progress over time with `NearbyMoveSelector` and the default configuration represented by `Late
Acceptance` with `</listChangeMoveSelector>` and `</listSwapMoveSelector>`.
On the graphs below you can see the
https://github.com/kiegroup/optaplanner/blob/main/optaplanner-examples/src/main/resources/org/optaplanner/examples/vehiclerouting/optional/benchmark/vehicleRoutingBenchmarkConfigListNearby.xml[optaplanner-benchmark]
 results for different amount of samples. You can find out more about optaplanner-benchmarks https://www.optaplanner.org/docs/optaplanner/latest/benchmarking-and-tweaking/benchmarking-and-tweaking.html[here].

== Execution time

> To analyze how much nearby reduces the

Reviewer: ```suggestion
To analyze how much nearby selection reduces the
```

> To analyze how much nearby reduces the

Reviewer: One sentence per line, please.
https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

Author: Fixed [d223fd8](https://github.com/kiegroup/optaplanner-website/pull/581/commits/d223fd82ab6cfb9c23e1a11ebec3fe75ea7fb69b)

time for finding the best solution lets compare the computational speed of the Late Acceptance (LA)
algorithm with the Nearby algorithm, and analyze how quickly each algorithm

> improves the solution score. The results indicate that the Nearby algorithm

Reviewer: I don't think it's a name of a well-known algorithm; thus I would not go for the capitals. More importantly, let's keep it consistent across the article.

```suggestion
improves the solution score. The results indicate that the nearby selection algorithm
```

Author: Added abbreviations
[d223fd8](https://github.com/kiegroup/optaplanner-website/pull/581/commits/d223fd82ab6cfb9c23e1a11ebec3fe75ea7fb69b)

consistently outperforms the LA algorithm in terms of speed of improvement.
In particular, you can see that the LV Nearby 20 variant was the fastest on average, while the Late Acceptance algorithm was the slowest.

image::scoreCalculationSpeedPerSecondNearby.png[How NearByMoveSelector improves score calculation speed]

[#table1]
.Score calculation count per second comparing Nearby and Default solvers
|===
| ^.^|Belgium-n50-k10
(2,890.00 samples)|Belgium-n100-k10 (10,790.00 samples)|Belgium-n500-k20
(258,980.00 samples)|Belgium-n1000-k20 (1,017,980.00 samples)
|Late Acceptance                      >|35,393.00/s >|37,334.00/s >|31,667.00/s >|30,282.00/s
|LV Nearby Parabolic Distribution 10  >|40,428.00/s >|45,059.00/s >|32,708.00/s >|32,036.00/s
|LV Nearby Parabolic Distribution 20  >|38,567.00/s >|40,552.00/s >|34,418.00/s >|37,188.00/s
|LV Nearby Parabolic Distribution 40  >|37,551.00/s >|39,833.00/s >|33,558.00/s >|35,125.00/s
|LV Nearby Parabolic Distribution 80  >|37,428.00/s >|35,625.00/s >|30,523.00/s >|36,060.00/s
|LV Nearby Parabolic Distribution 160 >|41,197.00/s >|38,086.00/s >|33,724.00/s >|34,451.00/s
|===

== Solution quality

> Despite the fact that Nearby improves the speed of calculation, it could potentially decrease the quality of the best solution.

Author: Added abbreviations [d223fd8](https://github.com/kiegroup/optaplanner-website/pull/581/commits/d223fd82ab6cfb9c23e1a11ebec3fe75ea7fb69b)

image::bestScoreNB.png[How NearByMoveSelector affect scores]

Here we see the score values for different solvers and datasets. You may see how much the nearby improves the score calculation while the dataset increases. From the other side look at the LV Nearby 10 bars. Despite being the fastest solution of all with the small dataset, it dramatically fell into local optimum and stopped improving the solution.

image::belgium-n500-k20BestScoreStatisticLevel.png[How local optimum affects the score]

== Scalability

image::bestScoreScalabilitySummaryLevel1.png[Scalability of Nearby solver]

For small datasets if the account of customers is no more than
500 it does not matter what to choose, but starting from the 250 000 samples
or 500 customers and 20 vehicles we can notice that the difference in the best

> scores achieved is almost 1.7X and the ratio increases to 2.63X reaching the 1m samples.

Reviewer: Is "1m" 1 million? Let's use the full word, please, abbreviations are often confusing.

> scores achieved is almost 1.7X and the ratio increases to 2.63X reaching the 1m samples.

Author: added description of graphs
[d223fd8](https://github.com/kiegroup/optaplanner-website/pull/581/commits/d223fd82ab6cfb9c23e1a11ebec3fe75ea7fb69b)

== Conclusion

> Nearby Move selector is a great chose for the problems that requires `@PlanningListVariable`. It speeds up score calculation and scales better that allows to tackle even more complex problems in the future.

Reviewer: ```suggestion
Nearby Move selector is a great choice for the problems that require `@PlanningListVariable`. It speeds up score calculation and scales better, which allows for tackling even more complex problems in the future.
```
